File: ros2_ws/src/robovision_ros2/robovision_4/src/robovision_client.cpp
#include "rclcpp/rclcpp.hpp"

#include "robovision_interfaces/srv/get_point_center.hpp"
#include "robovision_interfaces/msg/object_centroid.hpp"

#include <string>


class PointCloudCentroidNode : public rclcpp::Node
{
public:
    PointCloudCentroidNode() : Node("point_cloud_client"), point_({0.0, 0.0, 0.0})
    {
        //Declare parameters
        this->declare_parameter<int>("x", 320);
        this->declare_parameter<int>("y", 240);

        // Get parameter values
        x_ = this->get_parameter("x").as_int();
        y_ = this->get_parameter("y").as_int();

        point_.clear();

        // Processing timer
        //We call our service once each 2.5 seconds
        client_call_timer_ = this->create_wall_timer(
            std::chrono::milliseconds(2500),
            std::bind(&PointCloudCentroidNode::client_caller, this));

        RCLCPP_INFO(this->get_logger(), "Starting point_cloud_client application in cpp...");
    }

private:

    int x_, y_;
    std::vector<double> point_;

    rclcpp::TimerBase::SharedPtr client_call_timer_;
    std::vector<std::thread> threads_;

    void client_caller() 
    {
        try
        {
            get_point_center(x_, y_);
        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed.");
        }

        // Do something with the result
        if (!point_.empty()) 
        {
            RCLCPP_INFO(this->get_logger(),
                        "(%d, %d) position is: [%f, %f, %f]",
                        x_, y_, point_[0], point_[1], point_[2]);

            // Clear point data
            point_.clear();
        }
    }

    void get_point_center(int x, int y)
    {
        threads_.push_back(std::thread(std::bind(&PointCloudCentroidNode::call_get_point_center_server, this, x, y)));
    }

    void call_get_point_center_server(int x, int y) 
    {
        auto client = this->create_client<robovision_interfaces::srv::GetPointCenter>("get_point_center");
        while (!client->wait_for_service(std::chrono::seconds(1))) 
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for get_point_center service...");
        }

        auto request = std::make_shared<robovision_interfaces::srv::GetPointCenter::Request>();
        request->x = x;
        request->y = y;

        auto future = client->async_send_request(request);

        try
        {
            auto response = future.get();

            RCLCPP_INFO(this->get_logger(),
                        "(%d, %d) position is: [%f, %f, %f]",
                        x, y, response->point.centroid[0], response->point.centroid[1], response->point.centroid[2]);

            point_ = response->point.centroid;
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed.");
        }

    }

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PointCloudCentroidNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/install/robovision_services/lib/robovision_services/robovision_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial

from robovision_interfaces.msg import ObjectCentroid
from robovision_interfaces.srv import GetPointCenter

import numpy as np
import traceback


class PointCloudCentroidNode(Node):
    def __init__(self):
        super().__init__("point_cloud_client")

        #Declare parameters
        self.declare_parameter("x", 320)
        self.declare_parameter("y", 240)

        #Get parameter values
        self.x_ = self.get_parameter("x").value
        self.y_ = self.get_parameter("y").value

        self.point_ = []

        #Processing timer
        #We call our service once each 2.5 seconds
        self.client_call_timer_ = self.create_timer(2.5, self.client_caller)

        self.get_logger().info("Starting point_cloud_client application in python...")

    def client_caller(self):
        
        self.call_get_point_center_server(self.x_, self.y_)

        #Do something with the result
        if (self.point_):
            self.get_logger().info("(" + str(self.x_) + ", " + str(self.y_) + ") position is: " + str(self.point_))
            self.point_ = []

    #Basic server call template
    def call_get_point_center_server(self, _x, _y):
        client = self.create_client(GetPointCenter, "get_point_center")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for get_point_center service...")

        request = GetPointCenter.Request()
        request.x = _x
        request.y = _y

        future = client.call_async(request)

        #The callback only contains the response;
        #we explicity add the request for debugging purposes
        future.add_done_callback(
            partial(self.callback_call_point_cloud, _x=request.x, _y=request.y))
    
    def callback_call_point_cloud(self, future, _x, _y):
        try:
            response = future.result()
            self.get_logger().info("(" + str(_x) + ", " + str(_y) + ") position is: " + str(response.point))

            self.point_ = response.point
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))

def main(args=None):
    rclpy.init(args=args)
    node = PointCloudCentroidNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
